### PR TITLE
fix(#3884): useGrid: gracefully handle focus when a focused row is removed from the collection

### LINKED ISSUE #3884: useGrid: gracefully handle focus when a focused row is removed from the collection
# 🙋 Feature Request

Fix #3039 for useTable hook.

Removing a row from a Table as in the following storybook examples: 

1. [TableView: CRUD example](https://reactspectrum.blob.core.windows.net/reactspectrum/080488a859be32a1ee30f4f9fdcc4d5e8331a144/storybook/index.html?path=/story/tableview--crud)
2. [TableView: Inline delete buttons](https://reactspectrum.blob.core.windows.net/reactspectrum/080488a859be32a1ee30f4f9fdcc4d5e8331a144/storybook/index.html?path=/story/tableview--inline-delete-buttons)

## 🤔 Expected Behavior

Focus should move to the Actions menu button or inline Delete button in the next available row, or the previous available row when the last row is removed.

## 😯 Current Behavior

Focus moves to the first focusable element within the TableView FocusScope.

## 💁 Possible Solution

Use keyboard delegate to determine the keyBelow or keyAbove the focusedKey being removed, and set the focusedKey accordingly.

See examples with fix here:
1. [TableView: CRUD example (with possible fix)](https://reactspectrum.blob.core.windows.net/reactspectrum/69f979d9090ca659de9f7f6a127e230edaee3dca/storybook/index.html?path=/story/tableview--crud)
2. [TableView: Inline delete buttons  (with possible fix)](https://reactspectrum.blob.core.windows.net/reactspectrum/69f979d9090ca659de9f7f6a127e230edaee3dca/storybook/index.html?path=/story/tableview--inline-delete-buttons)
 
## 🔦 Context

Relates to #3039.

## 🧢 Your Company/Team

Adobe/Accessibility

### PR DESCRIPTION
**Pull request by majornista**
Closes #3884, relates to #3039

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issue/30).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

See examples with fix here:
1. [TableView: CRUD example (with possible fix)](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/tableview--crud)
2. [TableView: Inline delete buttons  (with possible fix)](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/tableview--inline-delete-buttons)
3. [ActionBar: default (with possible fix)](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/actionbar--default)
4. [ListView: Restore focus after item removal](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/listview-actions--restore-focus-after-item-removal)
5. [ListBox: Restore focus after deleting selected items](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/listbox--restore-focus-after-deleting-selected-items)

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-spectrum/table/stories/CRUDExample.tsx [hunk 1]
```diff
@@ -95,7 +95,11 @@ export function CRUDExample() {
             title="Delete"
             variant="destructive"
             primaryActionLabel="Delete"
-            onPrimaryAction={() => list.remove(dialog.item.id)}>
+            onPrimaryAction={() => {
+              // Add delay so that dialog closes and restores focus
+              // to the TableView before removing the row.
+              setTimeout(() => list.remove(dialog.item.id), 60);
```
Comment (snowystinger) [inline packages/@react-spectrum/table/stories/CRUDExample.tsx:101]: I don't think we're going to want this timeout
Where does focus go without the setTimeout? I think it goes to the first focusable in the scope? or does table move it to the first row?
If table does it, seems like we could consult the cached information we have an move it somewhere better
If the focus scope does it, then maybe the table can become the first focusable thing and when it gets focus, it then forwards it to the appropriate row?

Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx:101]: Without the timeout, focus moves to the first focusable in scope, the Select All checkbox.

Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx:101]: selectionManager.setFocusedKey executes within the useGrid and targets the appropriate key, but useRestoreFocus executes after, when the dialog closes.

File: packages/@react-aria/grid/src/useGrid.ts [hunk 2]
```diff
@@ -98,6 +98,41 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
...
+      state.selectionManager.focusedKey === null &&
+      cachedFocusedKey.current !== null
+    ) {
+      let keyBelow = cachedDelegate.current.getKeyBelow(cachedFocusedKey.current);
+      while (keyBelow) {
+        if (state.collection.getItem(keyBelow)) {
+          keyToFocus = keyBelow;
+          break;
+        }
+        keyBelow = cachedDelegate.current.getKeyBelow(keyBelow);
```
Comment (devongovett) [inline packages/@react-aria/grid/src/useGrid.ts:115]: I still kinda feel like maybe we should base this on the order in the new version of the collection rather than the old one? For example, if a row was removed and the data was re-sorted in a single transaction, this could produce strange results:

Starting state:

* A
* B   <-- focused
* C
* D
* E
* F

Ending state:

* D
* E
* A
* F
* C  <-- focused

If instead, we based it on the index of the item in the old collection, and simply focused the corresponding item at that same index in the new collection, it would end up like this:

* D
* E   <-- focused
* A
* F
* C

This might be more natural to the user. Otherwise they might be scrolled around a lot from where they were.

Seems like there could be a lot of real world cases like this, e.g. changing a filter. A bunch of new items could appear at the same time as some are removed, leaving focus in a completely different place.

Additionally, if we base this on the index, we wouldn't need the keyboard delegate to find the next key and could move this logic into a lower level layer shared among all collections. Basically:

```js
let node = lastCollection.getItem(focusedKey);
let newItem = [...newCollection][node.index];
setFocusedKey(newItem.key);
```

Might also need to skip over disabled items (depending on the `disabledBehavior`) also if the corresponding item is disabled.

Comment (majornista) [reply packages/@react-aria/grid/src/useGrid.ts:115]: We still need to worry about which column had focus in the row that was removed. Using the keyboard delegate makes this much simpler.

File: packages/@react-spectrum/table/stories/CRUDExample.tsx (hunk 1 above)
Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: I think I figured this out. The problem is that the FocusScope for Dialogs and Modals always `contain` focus using `useOverlayFocusContain`, so that when the user presses the Delete button to execute the primary action that removes the item, while the `useEffect` within `useGrid` does reset focus to the appropriate `focusedKey`, the FocusScope within the Overlay always steals focus back, even if the Overlay is in the process of closing. I think we need to account for the state of the Overlay when calling useOverlayFocusContain, so that an Overlay that is closing doesn't contain focus.

File: packages/@react-aria/grid/src/useGrid.ts (hunk 2 above)
Comment (devongovett) [reply packages/@react-aria/grid/src/useGrid.ts:115]: Ah yeah that's true. Using the index does seem promising though in terms of the behavior, would just need to be two dimensional for tables.

Comment (majornista) [reply packages/@react-aria/grid/src/useGrid.ts:115]: I'll investigate on Monday. Any thoughts on https://github.com/adobe/react-spectrum/pull/3885#discussion_r1063882356

File: packages/@react-spectrum/table/stories/CRUDExample.tsx (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: I think we were doing that at one point, when transitioning out. But I can't find it, maybe it was in some work we did but never merged.

What if, when focus is outside of a table or other collection, the children all became non-tabbable and the collection itself always received focus, then forwarded it wherever it was supposed to go?

Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: @snowystinger Is it this one? https://github.com/adobe/react-spectrum/pull/3704

Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: I can confirm that #3704 resolves the issue.

File: packages/@react-aria/grid/src/useGrid.ts [hunk 3]
```diff
@@ -98,6 +98,35 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
...
+  useEffect(() => {
+    if (
+      state.collection.size > 0 &&
+      state.selectionManager.focusedKey === null &&
+      cachedFocusedKey.current !== null
+    ) {
+      const node = cachedCollection.current.getItem(cachedFocusedKey.current);
+      const isCell = node.type === 'cell' || node.type === 'rowheader' || node.type === 'column';
+      const parentNode =
+        node.parentKey && isCell ?
```
Comment (devongovett) [inline packages/@react-aria/grid/src/useGrid.ts:112]: Keys can be falsy (e.g. `0`)

```suggestion
        node.parentKey != null && isCell ?
```

File: packages/@react-aria/grid/src/useGrid.ts [hunk 4]
```diff
@@ -98,6 +98,35 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
...
+        cachedCollection.current.getItem(node.parentKey) :
+        cachedCollection.current.getItem(node.key);
+      const rows = state.collection.rows;
+      const newRow = parentNode.index < rows.length ? rows[parentNode.index] : rows[rows.length - 1];
+      const keyToFocus =
+        newRow && newRow.hasChildNodes && isCell ?
+        [...newRow.childNodes][node.index].key :
+        newRow.key;
```
Comment (devongovett) [inline packages/@react-aria/grid/src/useGrid.ts:120]: We probably do need to skip over disabled rows in case the new row at that index can't be focused. It'll need to depend on the `disabledBehavior` since some items only disable selection and still allow focus - SelectionManager's `isDisabled` method should handle that.

File: packages/@react-aria/grid/src/useGrid.ts [hunk 5]
```diff
@@ -98,6 +98,35 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
...
+      if (keyToFocus) {
+        state.selectionManager.setFocusedKey(keyToFocus);
+      }
+    }
+    cachedFocusedKey.current = state.selectionManager.focusedKey;
+    cachedCollection.current = state.collection;
+  },
+  [state.collection, state.selectionManager, state.selectionManager.focusedKey]);
```
Comment (devongovett) [inline packages/@react-aria/grid/src/useGrid.ts:128]: Now that this doesn't depend on the keyboard delegate, maybe we can centralize this logic across components. Both `useGridState` and `useListState` already have some logic that sets the focused key to null if it gets deleted. Those should probably be replaced with this. Moving it to `useListState` would also mean that other components like `ListBox` would also get this behavior.

File: packages/@react-spectrum/table/stories/CRUDExample.tsx (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: No, it was elsewhere, the idea was that when a Popover began transitioning out, then contains would be set to false. I think it was this https://github.com/adobe/react-spectrum/pull/3725

Comment (majornista) [reply packages/@react-spectrum/table/stories/CRUDExample.tsx]: > No, it was elsewhere, the idea was that when a Popover began transitioning out, then contains would be set to false. I think it was this #3725

I tried incorporating the small changes in #3725, and it did not solve the problem.

File: packages/@react-aria/grid/src/useGrid.ts (hunk 5 above)
Comment (majornista) [reply packages/@react-aria/grid/src/useGrid.ts]: I have refactored the code into useGridState and useListState.

File: packages/@react-stately/grid/src/useGridState.ts [hunk 6]
```diff
@@ -46,17 +46,56 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
...
+            index = parentNode.index;
+          }
+          index--;
+        }
+      }
+      if (newRow) {
+        const keyToFocus =
+          newRow.hasChildNodes && parentNode !== node ?
+          [...newRow.childNodes][node.index].key :
+          newRow.key;
```
Comment (LFDanLu) [inline packages/@react-stately/grid/src/useGridState.ts:86]: What if the grid's column is deleted? I _think_ this might run into a problem since `node.index` won't exist in the new row's childNodes

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 7]
```diff
@@ -296,7 +297,8 @@ storiesOf('ListView/Actions', module)
           </Item>
         </ActionMenu>
       </>
-    ), args));
+    ), args))
+  .add('Restore focus after item removal', () => <FocusExample />);
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:301]:
```suggestion
  .add('Restore focus after item removal', (args) => <FocusExample {...args} />);
```
So we can test the difference in behavior between highlight and non-highlight selection

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 8]
```diff
@@ -2146,3 +2148,36 @@ function DragBetweenListsOverride(props) {
     </Flex>
   );
 }
+
+export const FocusExample = () => {
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:2152]:
```suggestion
export const FocusExample = (args) => {
```
and spread the args in line 2172

File: packages/@react-stately/list/src/useListState.ts [hunk 9]
```diff
@@ -51,17 +51,48 @@ export function useListState<T extends object>(props: ListProps<T>): ListState<T
...
+      const itemNodes = [...collection.getKeys()].map(
+        key => {
+          const itemNode = collection.getItem(key);
+          return itemNode.type === 'item' ? itemNode : null;
+        }
+      ).filter(node => node !== null);
+      let index = Math.min(startItem.index, itemNodes.length - 1);
+      let newNode:Node<T>;
+      while (index >= 0) {
+        if (!selectionManager.isDisabled(itemNodes[index].key)) {
```
Comment (LFDanLu) [inline packages/@react-stately/list/src/useListState.ts:73]: Same comment as in useGridState, prob should check for `selectionManager.disabledBehavior === 'all'` since focus behavior for disabled rows differs based on what is provided to ListView's `disabledBehavior` prop

**Review by LFDanLu — CHANGES_REQUESTED** [review]

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 10]
```diff
@@ -2146,3 +2148,36 @@ function DragBetweenListsOverride(props) {
...
+    {id: 10, name: 'Adobe Dreamweaver'}
+  ];
+
+  const list = useListData({
+    initialItems: items,
+    initialSelectedKeys: []
+  });
+
+  return (
+    <ListView aria-label="listview with removable items" width="250px" items={list.items} disabledKeys={['5']}>
```
Comment (majornista) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:2172]:
```suggestion
    <ListView aria-label="listview with removable items" width="250px" items={list.items} disabledKeys={['5']} {...args}>
```

File: packages/@react-spectrum/listbox/stories/ListBox.stories.tsx [hunk 11]
```diff
@@ -703,3 +709,57 @@ function App() {
     </>
   );
 }
+
+function FocusExample() {
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-spectrum/listbox/stories/ListBox.stories.tsx:713]: Indeed, I was taken aback by that at first as well. In selectable lists, items that can receive focus have an index, while section elements do not. If you remove a bunch of items at once, the index position might not align with your expectation. One alternative with which I could experiment would be to prioritize restoring focus to the previous index when the size difference between the cached items and the new items is greater than 1.

File: packages/@react-stately/grid/src/useGridState.ts (hunk 6 above)
Comment (majornista) [reply packages/@react-stately/grid/src/useGridState.ts:86]: Would this work to ensure that the `node.index` is present in the `newRow`?
```suggestion
        const keyToFocus =
          newRow.hasChildNodes  && parentNode !== node && node.index < newRow.childNodes.length  ?
          [...newRow.childNodes][node.index].key :
          newRow.key;
```

File: packages/@react-stately/grid/src/useGridState.ts [hunk 12]
```diff
@@ -46,17 +46,56 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
...
+      const node = cachedCollection.current.getItem(selectionState.focusedKey);
+      const parentNode =
+        node.parentKey != null && (node.type === 'cell' || node.type === 'rowheader' || node.type === 'column') ?
+        cachedCollection.current.getItem(node.parentKey) :
+        node;
+      const rows = collection.rows;
+      let index = Math.min(parentNode.index, rows.length - 1);
+      let newRow:GridNode<T>;
+      while (index >= 0) {
+        if (!selectionManager.isDisabled(rows[index].key)) {
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-stately/grid/src/useGridState.ts:67]: AFAIK, `selectionManager.isDisabled` checks `this.state.disabledBehavior === 'all'` at: https://github.com/adobe/react-spectrum/blob/93f49a443af04aba6ef2927f288eed340c788f5e/packages/%40react-stately/selection/src/SelectionManager.ts#L490-L492

File: packages/@react-stately/list/src/useListState.ts (hunk 9 above)
Comment (majornista) [reply packages/@react-stately/list/src/useListState.ts:73]: AFAIK, `selectionManager.isDisabled` checks `this.state.disabledBehavior === 'all'` at: https://github.com/adobe/react-spectrum/blob/93f49a443af04aba6ef2927f288eed340c788f5e/packages/%40react-stately/selection/src/SelectionManager.ts#L490-L492

File: packages/@react-stately/grid/src/useGridState.ts (hunk 6 above)
Comment (LFDanLu) [reply packages/@react-stately/grid/src/useGridState.ts]: At a glance I think so? I can try experimenting with a table whose columns/contents change after a settimeout

File: packages/@react-stately/list/src/useListState.ts (hunk 9 above)
Comment (LFDanLu) [reply packages/@react-stately/list/src/useListState.ts:73]: Ah, that is correct, this should be fine as is then

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 13]
```diff
@@ -296,7 +297,8 @@ storiesOf('ListView/Actions', module)
           </Item>
         </ActionMenu>
       </>
-    ), args));
+    ), args))
+  .add('Restore focus after item removal', (args = {}) => <FocusExample {...args} />);
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:301]: Mind adding one more parameter to line 66? Something like this:
```
    disabledBehavior: {
      control: {
        type: 'radio',
        options: ['selection', 'all']
      }
    }
```
a default can be provided to line 39 via `disabledBehavior: 'selection'`. I just realized I mistakenly conflated `highlight selection` and `disabledBehavior` when testing yesterday

File: packages/@react-spectrum/listbox/stories/ListBox.stories.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-spectrum/listbox/stories/ListBox.stories.tsx]: I'm fine with keeping this PR as is for now, would be good to see what other people think in testing, maybe split it out into a separate branch if need be? I _think_ your suggested experimental change would work, but I'd have to play around with it. Wish we had a way to differentiate between collection replacement and item deletion

File: packages/@react-stately/grid/src/useGridState.ts (hunk 6 above)
Comment (LFDanLu) [reply packages/@react-stately/grid/src/useGridState.ts]: Cool, verified this locally, seems to work well

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 14]
```diff
@@ -296,7 +304,9 @@ storiesOf('ListView/Actions', module)
           </Item>
         </ActionMenu>
       </>
-    ), args));
+    ), args))
+  .add('Restore focus after item removal (disabledBehavior: "selection")', (args) => <FocusExample {...args} />)
+  .add('Restore focus after item removal (disabledBehavior: "all")', (args) => <FocusExample {...args} disabledBehavior="all" />);
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:309]: Nit: we can remove this story since we have `disabledBehavior` as a storybook control.

**Review by LFDanLu — COMMENTED** [review]
Just a small nit and a test request, but otherwise looks good to me

File: packages/@react-stately/grid/src/useGridState.ts [hunk 15]
```diff
@@ -46,17 +46,59 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
...
+      }
+      if (newRow) {
+        const childNodes = newRow.hasChildNodes ? [...newRow.childNodes] : [];
+        const keyToFocus =
+          newRow.hasChildNodes &&
+          parentNode !== node &&
+          node.index < childNodes.length ?
+          childNodes[node.index].key :
+          newRow.key;
+        selectionState.setFocusedKey(keyToFocus);
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-stately/grid/src/useGridState.ts:90]: @LFDanLu For clarification, are you looking for a test of the HidingColumns story: https://reactspectrum.blob.core.windows.net/reactspectrum/7b34ac2366f2e3c2204a606d1afc012006540ddc/storybook/index.html?path=/story/tableview--hiding-columns&providerSwitcher-express=false , or some more elaborate example that removes both rows and columns at the same time?

[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-stately/grid/src/useGridState.ts:90]: I'd be fine with either, but leaning towards the latter. You could probably do something where you render something like
```
      <TableView aria-label="TableView with dynamic contents" width={300} height={200}>
        <TableHeader columns={columns}>
          {column => <Column>{column.name}</Column>}
        </TableHeader>
        <TableBody items={items}>
          {item =>
            (<Row key={item.foo}>
              {key => <Cell>{item[key]}</Cell>}
            </Row>)
          }
        </TableBody>
      </TableView>
```
in a test but with controlled items + columns or with items and columns provided via props. For the controlled items + columns route you could have a useEffect in said test component that sets a timeout on mount that replaces the items + columns. For the props route you could perform a `tree.rerender` with new columns/items. 

Either should let you then check the before/after focused item

File: packages/@react-spectrum/listbox/stories/ListBox.stories.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-spectrum/listbox/stories/ListBox.stories.tsx]: @LFDanLu See https://github.com/adobe/react-spectrum/commit/e0710b7b63e2dd990fb3a0eaf704066df08ec7d3

**Review by LFDanLu — DISMISSED** [review]
LGTM for the most part, lets get this tested by the rest of the team so we can compare the two methods of finding the new index to focus. One odd case that I found with the latest method was when I bulk selected and deleted "Snake", "Danni" and "Devon" from this [list](https://reactspectrum.blob.core.windows.net/reactspectrum/71c6ed79591e4229d763b6840856300eaecb8343/storybook/index.html?path=/story/listbox--restore-focus-after-deleting-selected-items&providerSwitcher-express=false). If the focusedKey was "Snake" then the newly focused key becomes "Aardvark" whereas I was expecting "Kangaroo", but otherwise it seems to be behaving well

File: packages/@react-stately/grid/src/useGridState.ts [hunk 16]
```diff
@@ -46,17 +46,67 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
...
+        node.parentKey != null && (node.type === 'cell' || node.type === 'rowheader' || node.type === 'column') ?
+        cachedCollection.current.getItem(node.parentKey) :
+        node;
+      const cachedRows = cachedCollection.current.rows;
+      const rows = collection.rows;
+      const diff = cachedRows.length - rows.length;
+      let index = Math.min(
+        (
+          diff > 1 ?
+          Math.max(parentNode.index - diff + 1, 0) :
```
Comment (devongovett) [inline packages/@react-stately/grid/src/useGridState.ts:69]: So if you deleted a row and at the same time added 2+ more the index would change but otherwise it stays the same? What was the reason for this?

Comment (majornista) [reply packages/@react-stately/grid/src/useGridState.ts:69]: Was added to try to address the concerns @LFDanLu raised with [the ListBox example](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/listbox--restore-focus-after-deleting-selected-items): https://github.com/adobe/react-spectrum/pull/3885#discussion_r1072940153.

When you remove multiple items, and depending on where focus is in relation to the items selected for removal, one perceives the focus jumping ahead significantly from where it was. This code intends to mitigate that perception. Do you think there is a better approach to this?

Comment (majornista) [reply packages/@react-stately/grid/src/useGridState.ts:69]: > So if you deleted a row and at the same time added 2+ more the index would change but otherwise it stays the same? What was the reason for this?

Also, in this use case the `diff` would be less than 1, because the new row count would be greater than the cached row count, `const diff = cachedRows.length - rows.length;`, so we would just use `Math.min(parentNode.index,  rows.length - 1)`.

Comment (devongovett) [reply packages/@react-stately/grid/src/useGridState.ts:69]: Ok that makes sense. One weird case I see is if you select two items from the bottom up so focus is on the first one, focus moves to the previous item rather than the next one. For example, select Snake and Kangaroo, delete, and focus ends on Aardvark rather than Danni (which is now at the corresponding index as Snake). If you select the opposite direction so focus is on Snake, then focus moves to Danni instead (rather than Devon which is at the corresponding index).

Comment (LFDanLu) [reply packages/@react-stately/grid/src/useGridState.ts:69]: I kinda like that behavior since it is restoring focus to the closest item in the direction the user's focus was closest/trending to if that makes sense (albeit that isn't what the code is explicitly doing). There is another odd case that I noted [here](https://github.com/adobe/react-spectrum/pull/3885#pullrequestreview-1266653110), so I think it would be good for the team to test this vs the previous implementation for opinions/edge cases and maybe we can make a call on if we are fine with this or if we want to refine some more

Comment (snowystinger) [reply packages/@react-stately/grid/src/useGridState.ts:69]: Mac Finder:
Select two, bottom first. Delete. Loses selection entirely.
Select two, top first. Delete. Selects item that replaced first selected index, not the focused one.
Select one. Delete. Selects the item that followed it.
Select two, top first, gap of n items between. Selects item that followed last focused+selected item.
Select two, bottom first, gap of n items between. Selects item that followed last focused+selected item.


Could improve on this by not losing selection in the first case and sending it to the next available item (checking more than one down in otherwords). The rest seems fine to use as the template for how to move?

**Review by devongovett — APPROVED** [review]
Approving for code. We can test and decide on final behavior for the edge cases.

**Review by LFDanLu — APPROVED** [review]
